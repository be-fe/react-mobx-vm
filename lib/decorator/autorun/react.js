'use strict';

exports.__esModule = true;

var _logger = require('../../utils/logger');

var _logger2 = _interopRequireDefault(_logger);

var _index = require('../utils/index');

var _core = require('./core');

var _core2 = _interopRequireDefault(_core);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable */
/**
 * @file   autorun
 * @author yucong02
 */
var keyMap = {
  initKey: 'componentDidMount',
  updateKey: 'componentWillReceiveProps',
  exitKey: 'componentWillUnmount'
};

exports.default = function () {
  for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  if ((0, _index.invokedWithArgs)(args)) {
    throw new Error('`autorun` don\'t need pass some arguments');
  }
  return _core2.default.apply(undefined, [keyMap].concat(args));
};