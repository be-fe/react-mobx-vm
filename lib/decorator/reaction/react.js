'use strict';

exports.__esModule = true;

var _core = require('./core');

var _core2 = _interopRequireDefault(_core);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = _core2.default.bind(null, {
  init: 'componentDidMount',
  exit: 'componentWillUnmount'
}); /**
     * @file: react
     * @author: Cuttle Cong
     * @date: 2018/2/15
     * @description:
     */