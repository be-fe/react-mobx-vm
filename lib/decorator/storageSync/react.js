'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

exports.default = function (urlKeyOrTarget, property, descriptor) {
  if ((0, _utils.invokedWithArgs)(arguments)) {
    var options = property;
    return function (target, property, descriptor) {
      return (0, _core2.default)(urlKeyOrTarget || property, (0, _extends3.default)({}, keyMap, options), target, property, descriptor);
    };
  }

  return (0, _core2.default)(property, keyMap, urlKeyOrTarget, property, descriptor);
};

var _core = require('./core');

var _core2 = _interopRequireDefault(_core);

var _utils = require('../utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file: index
 * @author: Cuttle Cong
 * @date: 2017/11/6
 * @description:
 */

var keyMap = {
  initKey: 'componentDidMount',
  updateKey: 'componentWillReceiveProps',
  exitKey: 'componentWillUnmount'
};