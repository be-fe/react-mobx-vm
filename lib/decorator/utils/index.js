'use strict';

exports.__esModule = true;

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

exports.invokedWithArgs = invokedWithArgs;
exports.newError = newError;
exports.invokedWithArgsForClass = invokedWithArgsForClass;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file   index
 * @author yucong02
 */

/** global: __DEVELOPMENT__ */

function invokedWithArgs(args) {
  return args.length !== 3 || (0, _typeof3.default)(args[0]) !== 'object' || typeof args[1] !== 'string' || (0, _typeof3.default)(args[2]) !== 'object';
}

function newError(type, message) {
  return new Error('@' + type + ': ' + message);
}

function invokedWithArgsForClass(args) {
  return args.length !== 1 || typeof args[0] !== 'function';
}