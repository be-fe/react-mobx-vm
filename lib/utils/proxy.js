'use strict';

exports.__esModule = true;
exports.default = proxy;

var _get = require('lodash/get');

var _get2 = _interopRequireDefault(_get);

var _set = require('lodash/set');

var _set2 = _interopRequireDefault(_set);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file: proxy
 * @author: Cuttle Cong
 * @date: 2018/1/23
 * @description:
 */
function proxy(host, path, getValue) {
  var old = (0, _get2.default)(host, path);
  var newVal = getValue(old);
  (0, _set2.default)(host, path, newVal);
  return newVal;
}