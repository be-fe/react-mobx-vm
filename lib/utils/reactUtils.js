'use strict';

exports.__esModule = true;
exports.isValidElement = undefined;
exports.isComponentInstance = isComponentInstance;
exports.isComponentClass = isComponentClass;
exports.assertReactClass = assertReactClass;
exports.convertReactElement = convertReactElement;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var isValidElement = exports.isValidElement = _react2.default.isValidElement; /**
                                                                               * @file: reactUtils
                                                                               * @author: Cuttle Cong
                                                                               * @date: 2018/2/6
                                                                               * @description:
                                                                               */
function isComponentInstance(instance) {
  return instance && instance instanceof _react2.default.Component;
}

function isComponentClass(com) {
  return com && isComponentInstance(com.prototype);
}

function assertReactClass(Component, message) {
  if (!isComponentClass(Component)) {
    throw new Error(message + ' require ReactClass');
  }
}

/**
 * 转换 ReactElem
 * @param element
 * @param rules: [[cond, process]]
 * @param parent
 * @return element
 */
function convertReactElement(element) {
  var rules = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
  var parent = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
  var outerIndex = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 0;

  if (!element || rules.length === 0) {
    return element;
  }

  function convert(element, index, children) {
    rules.forEach(function (_ref) {
      var cond = _ref[0],
          handle = _ref[1];

      if (cond(element, index, parent, children)) {
        var handledElem = void 0;
        // eslint-disable-next-line no-cond-assign
        if (typeof (handledElem = handle(element, index, parent, children)) !== 'undefined') {
          element = handledElem;
        }
      }
    });

    // @thinking return element ?
    return element;
  }

  if (Array.isArray(element)) {
    return _react2.default.Children.toArray(element).map(function (elem, index) {
      return convertReactElement(elem, rules, parent, index);
    });
  }
  var children = element && element.props && element.props.children;
  var newElement = element;
  newElement = convert(element, outerIndex, children);
  // convert may update children
  var newChildren = newElement && newElement.props && newElement.props.children;
  if (newChildren) {
    newChildren = convertReactElement(newChildren, rules, newElement, outerIndex);
  }

  if (newElement === element && newChildren === children) {
    return element;
  }

  return _react2.default.isValidElement(newElement) ? _react2.default.cloneElement(newElement, newElement.props, newChildren) : newElement;
}

exports.isElementOf = function (Component) {

  // Trying to solve the problem with 'children: XXX.isRequired'
  // (https://github.com/gaearon/react-hot-loader/issues/710). This does not work for me :(
  var originalPropTypes = Component.propTypes;
  Component.propTypes = void 0;

  // Well known workaround
  var elementType = _react2.default.createElement(Component).type;

  // Restore originalPropTypes
  Component.propTypes = originalPropTypes;

  return function (element) {
    return element && element.type === elementType;
  };
};