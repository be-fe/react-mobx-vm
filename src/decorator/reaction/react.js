/**
 * @file react
 * @author Cuttle Cong
 * @date 2018/2/15
 * @description
 */
import core from './core'

/** 
 * 发生改变之后，将会触发该成员方法
 * @name reactReaction
 * @public
 * @example
 * class View extends React.Component {
 *    \@observable val = '123'
 *    \@observable deep = { key: '123' }
 * 
 *    \@reaction('val', 'deep.key')
 *    reactionMethod(val, deepKey, dispose) {
 *      // 当 `this.val` `this.deep.key`
 *      // 发生改变之后，将会触发该方法
 *      // 与`autorun`不同的是，它第一次不会被触发
 *      
 *      // 调用 dispose 则会销毁掉该监听
 *      dispose && dispose()
 *    }
 * }
 */
export default core.bind(null, {
  init: 'componentDidMount',
  exit: 'componentWillUnmount'
})
