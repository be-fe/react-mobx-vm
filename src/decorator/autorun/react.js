/* eslint-disable */
/**
 * @file   autorun
 * @author yucong02
 */
import {
    invokedWithArgs
} from '../utils/index'
import autorun from './core'

const keyMap = {
    initKey: 'componentDidMount',
    updateKey: 'componentWillReceiveProps',
    exitKey: 'componentWillUnmount',
}

export default (...args) => {
    if (invokedWithArgs(args)) {
        console.warn('`autorun` 不需要传参')
        return autorun.bind(null, null)
    } else {
        return autorun(keyMap, ...args)
    }
}
