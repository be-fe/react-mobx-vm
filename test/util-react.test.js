/* eslint-disable react/react-in-jsx-scope */
/**
 * @file: util-react.test
 * @author: Cuttle Cong
 * @date: 2018/2/14
 * @description: $END$
 */
import * as React from 'react'
import { render } from 'enzyme'
import sinon from 'sinon'
import createReactClass from 'create-react-class'
import h from '../src/renderer/mixedRenderer'
import proxy from '../src/utils/proxy'
import * as util from '../src/utils/reactUtils'

class Component extends React.Component {
  render() {
    return <div>
      { this.props.name }
      { this.props.children }
    </div>
  }
}

describe('util-react', function () {
  test('util-react displayName', () => {
    class ABC extends React.Component {
      render() {
        return <div>
          { this.props.name }
          { this.props.children }
        </div>
      }
    }

    function CBA() {
    }


    expect(util.displayName(ABC)).toBe('ABC')
    expect(util.displayName(<ABC/>)).toBe('ABC')
    expect(util.displayName(<CBA/>)).toBe('CBA')
    expect(util.displayName(CBA)).toBe('CBA')
    expect(util.displayName(() => {})).toBe('Unknown')
  })

  test('util-react isComponentInstance', () => {
    expect(util.isComponentInstance(Component.prototype)).toBeTruthy()
  })

  test('util-react isComponentClass', () => {
    // eslint-disable-next-line react/display-name
    expect(util.isComponentClass(
      // eslint-disable-next-line react/display-name
      createReactClass({
        render: function() {
          return null
        }
      })
    )).toBeTruthy()
    expect(util.isComponentClass(<Component/>)).toBeFalsy()
    expect(util.isComponentClass(Component)).toBeTruthy()
  })

  test('util-react assertReactClass', () => {
    expect(
      () => util.assertReactClass(<Component/>, 'assertReactClass')
    ).toThrow('assertReactClass require ReactClass')
  })

  test('util-react convertReactElement', () => {
    const converted = util.convertReactElement(
      <div>
        <Component/>
      </div>, [
        [
          (element) => element.type === Component,
          (element) => React.cloneElement(element, { name: 'convert' }, <h1>Heading</h1>)
        ],
        [
          (element) => element.type === 'div',
          (element, index, parent, children) => React.cloneElement(element, { className: 'convert-classname' }, <h1>Big</h1>, children)
        ]
      ]
    )

    const dom = render(converted)
    expect(dom.text()).toBe('BigconvertHeading')
    expect(dom.attr('class')).toBe('convert-classname')
    expect(dom.html()).toMatchSnapshot()
  })

  test('util-react proxy', () => {
    const obj = {
      callback: () => {}
    }
    const spy = sinon.spy(obj, 'callback')
    proxy(obj, 'callback', function (callback) {
      return function () {
        callback && callback()
        return callback()
      }
    })

    obj.callback()
    expect(spy.callCount).toBe(2)
  })

  test('util-react isElementOf', () => {
    expect(util.isElementOf(Component)(<Component/>)).toBeTruthy()

    expect(util.isElementOf(React.Component)(<Component/>)).toBeFalsy()
  })
})
