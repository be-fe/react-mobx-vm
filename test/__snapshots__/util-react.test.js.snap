// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util-react util-react convertReactElement 1`] = `"<h1>Big</h1><div>convert<h1>Heading</h1></div>"`;
